### PR TITLE
Create parent classes for Gait, Joint, Limits and Setpoints

### PR DESCRIPTION
**Pull request by Roelemans**
Works on PM-124
Please advice me about naming.

## Description

Move much of the functionality from the Gait, Joint, Setpoint and Limits classes from the gait generator to parent classes, so that they can be used by other parts of the code (e.g. the gait selector).

## Changes

* Created a new package march_shared_classes
* Created parent classes for the Gait, Joint, Setpoint and Limits classes from the gait generator

### CODE REVIEW

File: march_shared_classes/package.xml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  <maintainer email="software@projectmarch.nl">Project March</maintainer>
+  <license>TODO</license>
+
+  <buildtool_depend>catkin</buildtool_depend>
+
+  <depend>march_shared_resources</depend>
+
+  <build_depend>roslint</build_depend>
+
+  <build_export_depend>robot_state_publisher</build_export_depend>
```
Comment (Olavhaasie) [inline march_shared_classes/package.xml]: this import is not used, so please remove it

File: march_shared_classes/src/march_shared_classes/gait/Gait.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            for setpoint in joint.setpoints:
+                timestamps.append(setpoint.time)
+
+        return sorted(set(timestamps))
+
+    def get_joint(self, name):
+        for i in range(0, len(self.joints)):
+            if self.joints[i].name == name:
+                return self.joints[i]
+        rospy.logerr("Joint with name " + name + " does not exist in gait " + self.name + ".")
```
Comment (Olavhaasie) [inline march_shared_classes/src/march_shared_classes/gait/Gait.py]:
```suggestion
        rospy.logerr('Joint with name ' + name + ' does not exist in gait ' + self.name)
```

File: march_shared_classes/src/march_shared_classes/gait/Gait.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+            joint_trajectory.joint_names.append(joint.name)
+
+        for timestamp in timestamps:
+            joint_trajectory_point = JointTrajectoryPoint()
+            joint_trajectory_point.time_from_start = rospy.Duration(timestamp)
+            for joint in self.joints:
+                interpolated_setpoint = joint.get_interpolated_setpoint(timestamp)
+
+                if interpolated_setpoint.time != timestamp:
+                    rospy.logerr("Time mismatch in joint " + joint.name + " at timestamp " + timestamp)
```
Comment (Olavhaasie) [inline march_shared_classes/src/march_shared_classes/gait/Gait.py]:
```suggestion
                    rospy.logerr('Time mismatch in joint ' + joint.name + ' at timestamp ' + timestamp)
```

File: march_shared_classes/src/march_shared_classes/gait/Gait.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+from trajectory_msgs.msg import JointTrajectoryPoint
+from march_shared_resources.msg import Setpoint
+
+
+class Gait:
+    def __init__(self, joints, duration, gait_type="walk_like",
+                 name="Walk", subgait="right_open", version="First try", description="Just a simple gait"):
+        # Set gait_type to walk_like if an old file with no gait_type is opened
+        if gait_type == "":
+            gait_type = "walk_like"
```
Comment (Olavhaasie) [inline march_shared_classes/src/march_shared_classes/gait/Gait.py]:
```suggestion
            gait_type = 'walk_like'
```

File: march_shared_classes/src/march_shared_classes/gait/Gait.py [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+from trajectory_msgs.msg import JointTrajectory
+from trajectory_msgs.msg import JointTrajectoryPoint
+from march_shared_resources.msg import Setpoint
+
+
+class Gait:
+    def __init__(self, joints, duration, gait_type="walk_like",
+                 name="Walk", subgait="right_open", version="First try", description="Just a simple gait"):
+        # Set gait_type to walk_like if an old file with no gait_type is opened
+        if gait_type == "":
```
Comment (Olavhaasie) [inline march_shared_classes/src/march_shared_classes/gait/Gait.py]:
```suggestion
        if gait_type == '':
```

File: march_shared_classes/src/march_shared_classes/gait/Setpoint.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
+class Setpoint:
+    digits = 4
```
[the remark this answers is not in the document]
Comment (Ishadijcks) [reply march_shared_classes/src/march_shared_classes/gait/Setpoint.py]: I hardcoded this number to indicate the amount of decimals that a setpoint should store.

File: march_shared_classes/src/march_shared_classes/gait/Gait.py (hunk 3 above)
Comment (Roelemans) [reply march_shared_classes/src/march_shared_classes/gait/Gait.py]: Though I tend to agree that these should be fixed, it is outside of the scope of this PR. Doing small fixes in a large refactor will make it difficult to oversee the changes.

File: march_shared_classes/src/march_shared_classes/gait/subgait.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
+import rospy
+from trajectory_msgs.msg import JointTrajectory
+from trajectory_msgs.msg import JointTrajectoryPoint
+from march_shared_resources.msg import Setpoint
+
+
+class Subgait:
+    def __init__(self, joints, duration, gait_type='walk_like',
+                 name='Walk', subgait='right_open', version='First try', description='Just a simple gait'):
```
Comment (Ishadijcks) [inline march_shared_classes/src/march_shared_classes/gait/subgait.py]: Do you think the attributes make sense if this is the class name?

Comment (Roelemans) [reply march_shared_classes/src/march_shared_classes/gait/subgait.py]: Good point. Renamed the name to gait_name and subgait to subgait_name.

**Pull request merged**